torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py

```diff
@@ -0,0 +1,149 @@
+    network. The idea is to attach the sparsifier to a layer (or layers) and it
+    zeroes out the activations based on the mask_fn (or sparsification function)
+    input by the user.
+    The mask_fn is applied once all the inputs are aggregated and reduced i.e.
+    mask = mask_fn(reduce_fn(aggregate_fn(activations)))
+
+    Note::
+        The sparsification mask is computed on the input **before it goes through the attached layer**.
+
+    Args:
+        model (nn.Module)
+            The model whose layers will be sparsified. The layers that needs to be
+            sparsified should be added separately using the register_layer() function
+
+        aggregate_fn (Optional, function type)
+            default aggregate_fn that is used if not specified while registering the layer.
+            specifies how inputs should be aggreagated over time.
+
+        reduce_fn (Optional, function type)
+            default reduce_fn that is used if not specified while registering the layer.
+            reduce_fn will be applied on the aggregated input
+
+        mask_fn (Optional, function type)
+            default mask_fn that is used to create the sparsification mask (based on the
+            aggregated & reduced input). This is used if not specified while registering the layer
+
+        features (Optional, list)
+            default selected features to sparsify.
+            If this is non-empty, then the mask_fn will be applied for each feature of the input.
+            For example,
+            >>> mask = [mask_fn(reduce_fn(aggregated_fn(input[feature])) for feature in features]
+
+        feature_dim (Optional, int)
+            default dimension of input features. Again, features along this dim will be chosen
+            for sparsification.
+
+        sparse_config (Dict)
+            Default configuration for the mask_fn. This config will be passed
+            with the mask_fn()
+
+    Expected Usage:
+        >>> model = SomeModel()
+        >>> act_sparsifier = ActivationSparsifier(...)  # init activation sparsifier
+
+        >>> # Initialize aggregate_fn
+        >>> def agg_fn(x, y):
+        >>>     return x + y
+
+        >>> # Initialize reduce_fn
+        >>> def reduce_fn(x):
+        >>>     return torch.mean(x, dim=0)
+
+        >>> # Initialize mask_fn
+        >>> def mask_fn(data):
+        >>>     return torch.eye(data.shape).to(data.device)
+
+
+        >>> act_sparsifier.register_layer(model.some_layer, aggregate_fn=agg_fn, reduce_fn=reduce_fn, mask_fn=mask_fn)
+
+        >>> # start training process
+            >>> # epoch starts
+                >>> # model.forward(), compute_loss() and model.backwards()
+            >>> # epoch ends
+            >>> act_sparsifier.step()
+        >>> # end training process
+
+        >>> sparsifier.squash_mask()
+    """
+    def __init__(self, model: nn.Module, aggregate_fn=None, reduce_fn=None, mask_fn=None,
```

Comment:
should this be passed in more along the lines of the normal sparsifier where a sparse defaults object is passed in?

Reply:
So, there are two kinds of config - 
1. function and feature level config: includes aggregate_fn, reduce_fn, mask_fn, features, etc. 
2. sparse config: Sparsity based config. These arguments are passed along with the mask_fn()

The normal sparsifier accepts sparse_config only where as in here we ask user to specify function, feature level config (explicit args) and sparse_config (similar to normal sparsifier).